_posts/2023-11-10-automated-test-generation-chess-puzzles-with-dafny.markdown

```diff
@@ -0,0 +1,288 @@
+
+In particular, running the command above results in the following two Dafny tests (formatted manually for this blog post). Don’t read too closely into the code just yet, we will visualize these tests soon. My point here is that Dafny produces tests written in Dafny itself and they can be translated to any of the languages the Dafny compiler supports.
+
+<div class="file" name="tests.dfy">
+{% highlight javascript linenos %}
+include "chess.dfy"
+module Tests {
+  import opened Chess
+  method {:test} Test0() {
+    var board := Board([Piece(kind:=King(White)  , at:=Pos(row:=6,col:=2)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=6,col:=7)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=5,col:=3)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=3,col:=6)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=5,col:=5))]);
+    Describe(board);
+  }
+  method {:test} Test1() {
+    var board := Board([Piece(kind:=King(White)  , at:=Pos(row:=0,col:=7)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=1,col:=5)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=1,col:=4)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=0,col:=5)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=0,col:=6))]);
+    Describe(board);
+  }
+}
+{% endhighlight %}
+</div><!--.file-->
+
+Note also that Dafny annotates the two test methods with the `{:test}` attribute. This means that after saving the tests to a file (`tests.dfy`), you can then use the `dafny test tests.dfy` command to compile and execute them. The default compilation target is C# but you can pick a different one using the `--target` option - read more about the `dafny-test` command [here](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-dafny-test).
+
+Oftentimes, you would want to translate the tests from Dafny code to some other format that you are using throughout your project. To do that, you can serialize every test by adding a set of print statements at the end of the test entry method. For our purposes, I wrote the `SerializeToSVG` method (included in the <a href="/blog/assets/src/chess.dfy">sources</a>) that converts each test to an `.svg` image of the corresponding chess board. You can download the relevant image assets <a href="/blog/assets/images/test-generation/assets">here</a> (originally from <a href="https://github.com/tabler/tabler-icons/tree/master">tabler icons</a>). Let us look at the results in picture form (knights are the pieces that look like a horse’s head and the white king is the only piece that isn’t filled in):
```

Comment:
> Oftentimes

Why not just "often" ?

Reply:
No particular reason) Have changed to "often"